cdip_connector/core/publisher.py

```diff
@@ -0,0 +1,74 @@
+import json
+import logging
+from abc import ABC, abstractmethod
+
+from confluent_kafka import Producer
+
+from cdip_connector.core import cdip_settings
+
+logger = logging.getLogger(__name__)
+
+
+class Publisher(ABC):
+
+    @abstractmethod
+    def publish(self, topic: str, data: dict, extra: dict = {}):
+        ...
+
+
+class KafkaPublisher(Publisher):
+
+    def __init__(self):
+        cloud_enabled = cdip_settings.CONFLUENT_CLOUD_ENABLED
+
+        config_dict = {'bootstrap.servers': cdip_settings.KAFKA_BROKER}
+
+        if cloud_enabled:
+            config_dict['security.protocol'] = 'SASL_SSL'
+            config_dict['sasl.mechanisms'] = 'PLAIN'
+            config_dict['sasl.username'] = cdip_settings.CONFLUENT_CLOUD_USERNAME
+            config_dict['sasl.password'] = cdip_settings.CONFLUENT_CLOUD_PASSWORD
+
+        self.producer = Producer(config_dict)
+
+    @staticmethod
+    def create_message_key(data):
+        integration_id = data.get('integration_id')
+        device_id = data.get('device_id')
+        if integration_id and device_id:
+            return f'{integration_id}.{device_id}'
+        else:
+            # logger.warning(f'Unable to determine key, integration_id or device_id not present in observation')
+            return None
+
+    def publish(self, topic: str, data: dict, extra: dict = {}):
```

Comment:
It's much safer to use immutable objects for default values. I recommend doing something like this:
```
def publish(self, topic: str, data: dict, extra: dict = None):
    extra = extra or {}
    ...
```